update/v1alpha1/types_update_status_progress.go

```diff
@@ -0,0 +1,424 @@
+
+	// estimatedCompletedAt is the estimated time when the update will complete
+	// +optional
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	EstimatedCompletedAt *metav1.Time `json:"estimatedCompletedAt,omitempty"`
+}
+
+// ControlPlaneAssessment is the assessment of the control plane update process
+// +kubebuilder:validation:Enum=Unknown;Progressing;Completed;Degraded
+type ControlPlaneAssessment string
+
+const (
+	// Unknown means the update status and health cannot be determined
+	ControlPlaneAssessmentUnknown ControlPlaneAssessment = "Unknown"
+	// Progressing means the control plane is updating and no problems or slowness are detected
+	ControlPlaneAssessmentProgressing ControlPlaneAssessment = "Progressing"
+	// Completed means the control plane successfully completed updating and no problems are detected
+	ControlPlaneAssessmentCompleted ControlPlaneAssessment = "Completed"
+	// Degraded means the process of updating the control plane suffers from an observed problem
+	ControlPlaneAssessmentDegraded ControlPlaneAssessment = "Degraded"
+)
+
+// ControlPlaneUpdateVersions contains the original and target versions of the upgrade
+type ControlPlaneUpdateVersions struct {
+	// previous is the version of the control plane before the update. When the cluster is being installed
+	// for the first time, the version will have a placeholder value '<none>' and carry 'Installation' metadata
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.version == '<none>' ? (has(self.metadata) && self.metadata.exists(m, m.key == 'Installation')) : !(has(self.metadata) && self.metadata.exists(m, m.key == 'Installation'))",message="previous version must be '<none>' iff marked with Installation metadata"
+	Previous Version `json:"previous"`
+
+	// target is the version of the control plane after the update. It may never be '<none>' or have `Installation` metadata
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.version != '<none>' && !(has(self.metadata) && self.metadata.exists(m, m.key == 'Installation'))",message="target version must not be '<none>' or have Installation metadata"
+	Target Version `json:"target"`
+}
+
+// Version describes a version involved in an update, typically on one side of an update edge
+type Version struct {
+	// version is a semantic version string, or a placeholder '<none>' for the special case where this
+	// is a "previous" version in a new installation, in which case the metadata must contain an item
+	// with key 'Installation'
+	// +required
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:MinLength=5
+	// +kubebuilder:validation:MaxLength=64
+	Version string `json:"version"`
+
+	// metadata is a list of metadata associated with the version. It is a list of key-value pairs. The value is optional
+	// and when not provided, the metadata item has boolean semantics (presence indicates true)
```

Comment:
What does this mean? Can you provide an example?

Reply:
It's the same as with kube labels - sometimes you only care about "has / does not have a label", not about its actual value.

For version I am now aware of three metadata use cases, two of them are boolean in semantics, one needs value:

1. version is empty because the cluter was installed this way and never updated
2. previous version is a "partial" one - the current update was triggered in the middle of the update to that version, so previous version was never fully completed
3. architecture of a version (=of the payload image). there is a special case where you can do a same version update, from a single arch to multi

Reply:
Right I see what you mean, I know this API might evolve so lets hold off too much bike shedding here, but I think we may want to think about the wording here to make this clearer